### PR TITLE
Teach the requirements command to also list plugins

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt [hunk 1]
```diff
@@ -141,8 +148,27 @@ class RequirementsCommand : OrtCommand(
...
     }
 
-    private fun getToolsByCategory(): Map<String, List<CommandLineTool>> {
-        val reflections = Reflections("org.ossreviewtoolkit", Scanners.SubTypes)
+    private fun listPlugins(reflections: Reflections) {
+        val pluginClasses = reflections.getSubTypesOf(Plugin::class.java)
+
+        val pluginTypes = pluginClasses.mapNotNull { clazz ->
+            val companion = clazz.declaredClasses.find { it.name.endsWith("\$Companion") }
+            val all = runCatching { companion?.getDeclaredMethod("getALL") }.getOrNull()
```
Comment (fviernau) [inline plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:167]: As this line is relatively easy to break, shall we add a test which verifies the at least one plugin is listed?

Also is "getALL" actually right, or should it be "getAll"?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt [hunk 2]
```diff
@@ -51,7 +54,11 @@ class RequirementsCommand : OrtCommand(
     help = "Check for the command line tools required by ORT."
 ) {
     override fun run() {
-        val statusCode = checkToolVersions()
+        val reflections = Reflections("org.ossreviewtoolkit", Scanners.SubTypes)
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:68]: You mean to move `reflections` to the class level?

[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:68]: Yes.

[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:68]: Ok, done.

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt [hunk 3]
```diff
@@ -141,8 +148,27 @@ class RequirementsCommand : OrtCommand(
...
+        val pluginClasses = reflections.getSubTypesOf(Plugin::class.java)
+
+        val pluginTypes = pluginClasses.mapNotNull { clazz ->
+            val companion = clazz.declaredClasses.find { it.name.endsWith("\$Companion") }
+            val all = runCatching { companion?.getDeclaredMethod("getALL") }.getOrNull()
+
+            all?.let {
+                @Suppress("UNCHECKED_CAST")
+                val plugins = all.invoke(clazz.kotlin.companionObjectInstance) as Map<String, *>
+                clazz.simpleName to plugins.keys
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:172]: `clazz` is not an instance of a `Plugin`, but of type `Class<Plugin>`, so I cannot access `Plugin.type` here as no plugin is instantiated.

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt [hunk 4]
```diff
@@ -53,12 +57,19 @@ class RequirementsCommand : OrtCommand(
     name = "requirements",
     help = "Check for the command line tools required by ORT."
 ) {
+    private enum class RequirementsType { PLUGINS, COMMANDS }
+
+    private val list by option(
+        "--list", "-l",
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:63]: I believe "--list plugins" reads nicer than "--types plugins" (in the sense of a human-readable sentence with a verb).

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt [hunk 5]
```diff
@@ -53,12 +57,19 @@ class RequirementsCommand : OrtCommand(
...
+        help = "A comma-separated list of requirements to list."
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:64]: For reasons related to [my comment below](https://github.com/oss-review-toolkit/ort/pull/7935#discussion_r1405808633), I'd prefer to not coin the term "requirements type" here.

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt (hunk 1 above)
Comment (sschuberth) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:167]: > Also is "getALL" actually right, or should it be "getAll"?

It's correct, as it's Java syntax to refer to Kotlin's `ALL` property.

File: plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/commands/requirements/src/main/kotlin/RequirementsCommand.kt:68]: Forgot to commit / push?

**Review by fviernau — APPROVED** [review]

**Pull request merged**